prov/shm: add read protocol if CMA is disabled

aingerson opened the pull request:
Fixes #5656 
@rajachan @zhngaj @shefty This is a rough draft. I'm working on adding the large message segmentation on top of this which will use a similar approach but I'm opening this up for review and testing now. Let me know your initial thoughts and if this fixes the issue you are seeing with RMA reads.

zhngaj commented:
bot:aws:retest

zhngaj commented on prov/shm/src/smr_atomic.c:
```diff
@@ -119,6 +119,45 @@ static void smr_format_inject_atomic(struct smr_cmd *cmd, fi_addr_t peer_id,
+					      compv, comp_count, 0);
+		if (comp_size != cmd->msg.hdr.size)
+			FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+				"atomic and compare buffer size mimatch\n");
```
NIT: mismatch

shefty commented on prov/shm/src/smr.h:
```diff
@@ -225,7 +233,15 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+		void *context, struct smr_region *smr,
+		struct smr_region *peer_smr, struct smr_inject_buf **tx_bufs,
+		uint64_t buf_count, struct smr_resp *resp,
+		struct smr_tx_entry *pend);
```
This is an impressive number of arguments. Maybe it would be cleaner to make use of 2-3 format calls that the caller could invoke?

shefty commented on prov/shm/src/smr_ep.c:
```diff
@@ -303,17 +308,46 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
 	cmd->msg.hdr.size = total_len;
+	memcpy(cmd->msg.data.iov, iov, sizeof(*iov) * count);
+
+	smr_post_pend_resp(pend, cmd, context, iov, count, resp);
```
see below

shefty commented on prov/shm/src/smr_ep.c:
```diff
@@ -303,17 +308,46 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
+		cmd->msg.data.offset[i] = (uintptr_t)
+					((char **) tx_bufs[i] - (char **) peer_smr);
+
+	smr_post_pend_resp(pend, cmd, context, iov, count, resp);
```
Please restrict what the format calls do.  They should only prepare/format a structure.  The posting of the send should come outside of this function, otherwise, it looks like a side effect and is not expected by the function name.

zhngaj replied:
I agree that we should restrict what the format calls do. 
I think `smr_post_pend_resp` is not doing `real` post. It just does a few assignments though, associating `resp` with `pend`, and `pend` with `cmd`.

shefty commented on include/ofi_shm.h:
```diff
@@ -72,6 +72,7 @@ enum {
 	smr_src_inline,	/* command data */
 	smr_src_inject,	/* inject buffers */
 	smr_src_iov,	/* reference iovec via CMA */
+	smr_src_offset,	/* use inject buffer offsets with resp*/
```
Maybe call this SAR instead of offset?  The use of the term offset isn't clear to me.  If I understand the behavior, this is using multiple inject buffers.

aingerson commented on include/ofi_shm.h:
```diff
@@ -110,6 +111,7 @@ struct smr_msg_hdr {
 
 #define SMR_MSG_DATA_LEN	(128 - sizeof(struct smr_msg_hdr))
 #define SMR_COMP_DATA_LEN	(SMR_MSG_DATA_LEN / 2)
+#define SMR_MAX_OFFSET_CNT	((SMR_MSG_DATA_LEN - 8) / 8)
```
Yes, which technically should only need 1 byte, but I rounded to align the data. This is also what the iov numbers get calculated which would also be 1 for the iov_count. I don't know if this alignment is actually an issue though - probably best to discuss with @shefty

shefty replied:
I would keep 64-bit values aligned.

zhngaj commented on prov/shm/src/smr_ep.c:
```diff
@@ -242,10 +242,15 @@ static void smr_init_queue(struct smr_queue *queue,
+	pend->context = context;
+	memcpy(pend->iov, iov, sizeof(*iov) * iov_count);
+	pend->iov_count = iov_count;
+	
```
NIT: There's a tab space in this line.

aingerson commented:
> Let's see if I have this correct. The initiator is performing an RMA read operation. It reserves several tx buffers to use as target bounce buffers for the peer to write into. Correct?...

In this instance, yes. The aim is to have this same protocol be the same one that the large message segmentation use. So in regards to your "counter intuitive" comment, you're right but also worth noting that this is not the primary use case. It just happens that the RMA read target side is using a tx_buf. Also worth noting that the tx_buf is still being managed/used by the initiator for the purposes of sending data so it sort of isn't counter intuitive...?

Regardless, I agree to rename the protocol to SAR to help with understanding.

Creating a separate pool of buffers would help but, obviously, there would a significant memory increase. Thoughts on the trade off?

shefty commented:
The memory increase depends on the buffer size and number that can be allocated.  We could limit the number of large transfers to a handful, or reduce the number of inject buffers accordingly.  As an example, we could allocated 2 SAR buffers per large transfer, with a limit of 4-8 outstanding SAR transfers at a time.

A separate pool could allow small transfers to proceed, without being blocked behind a large SAR transfer.  Does the implementation allow a large transfer to consume all tx buffers?  If so, transfers to other peers may be impacted while there is a large transfer outstanding to one.

aingerson commented:
> A separate pool could allow small transfers to proceed, without being blocked behind a large SAR transfer. Does the implementation allow a large transfer to consume all tx buffers? If so, transfers to other peers may be impacted while there is a large transfer outstanding to one.
The implementation maxes out large transfers to consume no more than 8 inject buffers each but doesn't limit how many SAR transfers can be outstanding at once. We could limit the number of outstanding SAR transfers to help not consume all the buffers. But then I guess we might as well have a separate pool for SAR buffers to simplify. We could allocate 2 SAR buffers per peer (right now the maximum number of peers shm can handle is hard coded at 256) and dedicate those 2 buffers to that peer so we don't prevent other peers from making progress when one is trying to send a large SAR message.

rajachan commented:
@aingerson We would like to get Jie's PR (https://github.com/ofiwg/libfabric/pull/5688) in for 1.10 so we have a good fallback story in CMA-lacking environments. Curious what your plans are for this PR.

(Tracking the other EFA PRs we'd like to get in for the 1.10 release in [this](https://github.com/ofiwg/libfabric/milestone/21) milestone)

aingerson commented:
@rajachan I was hoping to get this into 1.10 as well. This includes the base changes for the large message segmentation as well so I'm currently working on finalizing those and including them here. I'll try to get something up soon so we can compare the 2 PRs and make sure we're on the same page.

aingerson commented:
@shefty @rajachan @zhngaj 
Updated. I've added a few extra fixes and refactoring/reformatting/simplifying as well.
I've used a different approach for the rma reads and am removing the large message segmentation from this PR. I still have a set of changes in the works for it, but it was starting to get out of hand I want to make sure we have this fix in for 1.10.

aingerson commented on prov/shm/src/smr_atomic.c:
```diff
@@ -232,31 +178,47 @@ static ssize_t smr_generic_atomic(struct smr_ep *ep,
 			assert(ioc);
 			ofi_ioc_to_iov(ioc, iov, count, ofi_datatype_size(datatype));
 		} else {
+			flags |= SMR_RMA_REQ;
```
Going to remove this in the next revision, but holding off to not overload the CI.
Atomic reads should always go through the fetch case so it should already be set. All ops fall through to here also so this is getting set for all ops, incorrectly.

aingerson commented:
Any thoughts on this before I update again?

shefty commented:
Can you pull out the first 3 patches into another PR, so those can be merged immediately?

shefty commented on prov/shm/src/smr_av.c:
```diff
@@ -154,8 +154,8 @@ static int smr_av_lookup(struct fid_av *av, fi_addr_t fi_addr, void *addr,
-	((char *) addr)[*addrlen] = '\0';
-	*addrlen = sizeof(struct smr_addr);
+	((char *) addr)[MIN(*addrlen - 1, strlen(smr_name(peer_smr)))] = '\0';
+	*addrlen = strlen(smr_name(addr) + 1);
```
Is this needed for v1.9.x?

aingerson replied:
Yes, I'll cherry pick

shefty commented on prov/shm/src/smr_atomic.c:
```diff
@@ -196,12 +192,14 @@ static ssize_t smr_generic_atomic(struct smr_ep *ep,
 		return ret;
 
-	peer_smr = smr_peer_region(ep->region, peer_id);
+	peer_smr = smr_peer_region(ep->region, id);
```
Is this bug fix needed in v1.9.x?

aingerson replied:
no, this is just refactoring

shefty replied:
I don't know if this is the bug fix.  I'm referring to the bug fix in the commit message, which should be pulled out into its own commit.

aingerson replied:
Gotcha, will pull that bug fix out and include in the new PR

aingerson commented:
> Can you pull out the first 3 patches into another PR, so those can be merged immediately?

How do you feel about pulling the first 5 out? Ie
(1) ep_entry->rx_entry + add tx_entry
(2) align cmd data
(3) return correct addrlen
(4) fix typo
(5) refactor format functions
so @zhngaj can rebase and update the mmap changes with the refactored format calls?

shefty commented:
I'm in favor of pulling out any that can be merged now.

shefty commented on prov/shm/src/smr_atomic.c:
```diff
@@ -116,53 +116,6 @@ static void smr_format_inject_atomic(struct smr_cmd *cmd,
-	smr_format_pend_resp(pend, cmd, NULL, result_iov, count, resp);
-	ofi_cirque_commit(smr_resp_queue(ep->region));
-}
-
```
Please update the commit title to indicate that this is fixing a potential data corruption/ordering issue, so that it's clear we may want to backport it as a bug fix.  And on that note, do we?

aingerson replied:
Ah sorry just saw this comment. Probably should back port this fix. I'll open a PR.

aingerson commented:
See #5762 - pulled out most of these commits. Waiting to update this PR until those are merged to rebase

aingerson commented:
Status on this? Any more comments? @shefty

shefty commented:
I'm happy.  :)  <- see?
